src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,16 +417,365 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+                    internalRequest.PathHandler,
+                    new List<ODataPathSegment>());
+                ODataPath odataPath = pathHandler.Parse(serviceRoot, id, internalRequest.RequestContainer).Path;
+
+                return odataPath;
+            }
+            catch (Exception)
+            {
+                return null;
+            }
+        }
+
+        private static void ApplyODataIdContainer(object resource, ODataResourceWrapper resourceWrapper,
+            ODataDeserializerContext readContext)
+        {
+            Routing.ODataPath path = readContext.Path;
+            if (path == null)
+            {
+                return;
+            }
+
+            IEdmEntityType entityType = path.EdmType.AsElementType() as IEdmEntityType;
+
+            if (entityType != null)
+            {
+                //Setting Odataid , for POCO classes, as a property in the POCO object itself(if user has OdataIDContainer property),
+                //for Delta and EdmEntity object setting as an added property ODataIdcontianer in those classes
+
+                ODataPath ODataPath = new ODataPath(path.Segments);
+
+                // if there is no Id on the resource, try to compute one from path
+                if (resourceWrapper.ResourceBase.Id == null)
+                {
+                    ODataUri odataUri = new ODataUri { Path = ODataPath };
+                    resourceWrapper.ResourceBase.Id = odataUri.BuildUri(ODataUrlKeyDelimiter.Parentheses);
+                }
+
+                if (resourceWrapper.ResourceBase.Id != null)
+                {
+                    string odataId = resourceWrapper.ResourceBase.Id.OriginalString;
+
+                    IODataIdContainer container = new ODataIdContainer();
+                    container.ODataId = odataId;
+
+                    if (resource is EdmEntityObject edmObject)
+                    {
+                        edmObject.ODataIdContainer = container;
+                        edmObject.ODataPath = ODataPath;
+                    }
+                    else if (resource is IDeltaSetItem deltasetItem)
+                    {
+                        deltasetItem.ODataIdContainer = container;
+                        deltasetItem.ODataPath = ODataPath;
+                    }
+                    else
+                    {
+                        PropertyInfo containerPropertyInfo = EdmLibHelpers.GetClrType(entityType, readContext.Model).GetProperties().Where(x => x.PropertyType == typeof(IODataIdContainer)).FirstOrDefault();
+                        if (containerPropertyInfo != null)
+                        {
+                            IODataIdContainer resourceContainer = containerPropertyInfo.GetValue(resource) as IODataIdContainer;
+                            if (resourceContainer != null)
+                            {
+                                containerPropertyInfo.SetValue(resource, resourceContainer);
```

Comment:
We retrieved this value from `resource`, i.e.
```csharp
IODataIdContainer resourceContainer = containerPropertyInfo.GetValue(resource) as IODataIdContainer;
```
What is the value in setting it here?
```csharp
containerPropertyInfo.SetValue(resource, resourceContainer);
```

Reply:
So we get the property from the POCO class, then we set the `ODataId` value to the property. The first one is just the property that was defined in the POCO class. The second statement we are assigning the ODataId value that has been deserialized to the `IODataIdContainer`